src/proposals/proposal.cpp

```diff
@@ -131,18 +153,66 @@ namespace hypha
         }
     }
 
+    void Proposal::publish(const eosio::name &proposer, Document &proposal)
+    {
+        TRACE_FUNCTION()
+        eosio::checksum256 root = getRoot(m_dao.get_self());
+        eosio::checksum256 memberHash = Member::calcHash(proposer);
+
+        EOS_CHECK(Edge::exists(m_dao.get_self(), root, proposal.getHash(), STAGING_PROPOSAL), "Only proposes in staging can be published");
+        EOS_CHECK(Edge::exists(m_dao.get_self(), memberHash, proposal.getHash(), common::OWNS), "Only the proposer can publish the proposal");
+        Edge::get(m_dao.get_self(), root, proposal.getHash(), STAGING_PROPOSAL).erase();
+        _publish(proposer, proposal, root);
+    }
+
+    void Proposal::remove(const eosio::name &proposer, Document &proposal)
+    {
+        TRACE_FUNCTION()
+        eosio::checksum256 root = getRoot(m_dao.get_self());
+        eosio::checksum256 memberHash = Member::calcHash(proposer);
+
+        EOS_CHECK(Edge::exists(m_dao.get_self(), root, proposal.getHash(), STAGING_PROPOSAL), "Only proposes in staging can be removed");
+        EOS_CHECK(Edge::exists(m_dao.get_self(), memberHash, proposal.getHash(), common::OWNS), "Only the proposer can remove the proposal");
+
+        m_dao.getGraph().eraseDocument(proposal.getHash(), true);
+
+        eosio::action(
+            eosio::permission_level{this->m_dao.get_self(), name("active")},
+            COMMENTS_CONTRACT, name("delsection"),
+            std::make_tuple(
+                this->m_dao.get_self(), // scope. todo: use tenant here
+                proposal.getContentWrapper().getOrFail(SYSTEM, COMMENT_NAME, "Proposal has no comment section")->getAs<eosio::name>() // section
```

Comment:
Would this not cause a failure of the action when there are no comments? But, documents are allowed to have no comments, right?

Reply:
I'll check this - They might not have comments but could have the (empty) section. Could probably be optimized to only create the section on first comment.